### ISSUE TITLE
EAMxx: intelligently handle fillval tracking in diags

### ISSUE THREAD

**Issue by mahf708** [E3SM-Project/E3SM#6803]
Diagnostics being their own derived classes are supposed to handle variety of inputs/outputs. 

- Sometimes diagnostics can produce their own fillval depending on conditions. Take FieldAtXyz being unavailable on a mountain or something. Another is OpticalDepth not being available at night.
- Other time, diagnostics must take into them masked fields, and thus will have to reckon with these masked/fill values.

In both cases, diagnostics are supposed to be able to cleanly and intelligently handle accumulation. Currently, only the FieldAtXyz is handled okay. 

Options:
1. Make all diagnostics redundantly track accumulation. This is not desirable as it is a blanket solution
2. Introduce a functionality to determine if a diagnostic needs to track the accumulation carefully based on the conditions above (i.e., it either takes in a field that requires careful handling or that it itself produces fillvals thus needing careful handling)

**Comment by bartgol** [comment]
What do you mean with "track accumulation"? Do you mean track averaging tally? I don't think diags are the right place to handle averaging. An output stream may have a mix of FM fields and diagnostics, so having diags do the tally while FM fields don't would make IO even more complex. Besides, the avg count is the same for all "field_xyz_at_1000hPa", so IO is able to only allocate _one_ field for the avg counting.

I think we can make diags tell IO whether they may generate mask/fill values. Some diags may do it directly (like FieldAtHeight), while others may do it indirectly. E.g., in `aodvis_at_modeltop`, the "outer" diag is FieldAtLevel, which does not per se generate Fill/Mask values, but the overall output may, since the inner diag does generate them.

I propose adding to `AtmosphereDiagnostic` an interface like
```c++
virtual void can_generate_invalid_values () const { return false; }
```
Being false by default, all diags that do not generate invalid values don't need to be modified. But FieldAtHeight or AodVis can return true. IO will then take care of checking if a diag can generate invalid values "indirectly". E.g., like this
```c++
   bool can_gen_fill_val = diag->can_generate_invalid_values();
   for (const auto& dep : m_diag_depends_on_diag[diag->name()]) {
      can_gen_fill_val &= m_diagnostics.at(dep)->can_generate_invalid_values()
   }
```

**Comment by bartgol** [comment]
> > Being false by default, all diags that do not generate invalid values don't need to be modified. But FieldAtHeight or AodVis can return true. IO will then take care of checking if a diag can generate invalid values "indirectly". E.g., like this
> 
> Yes, but we also need to handle the stuff it's receiving ... maybe:
> 
> ```c++
> virtual void can_generate_invalid_values () const { 
>   bool some_bool;
>   for (const auto& f : get_fields_in()){
>     some_bool ||= f.has_mask();
>   }
>   return some_bool;
> }
> ```
> 
> then override inside diag if we are sure it may make invalid values itself, etc.

I don't think so. If an input field X is from the FM, then it's out of the diag responsibility. It is no different than outputing X directly. If we want to handle FM fields that MAY be masked/filled, then we need to also do that check on the metadata for every single field in the Output class...

**Comment by meng630** [comment]
Thanks @mahf708 for bringing this to our attention! I'll keep an eye on this issue.

**Comment by susburrows** [comment]
Hi @mahf708 , would you be able to revisit this issue soon?  We'll need to AOD diagnostic output to be ready soon for evaluation of the ne256 simulations.

**Comment by mahf708** [comment]
@bartgol, let's prioritize fixing this? I am not sure where your thinking is, but we are reaching a point where we likely want any solution. Either brute force or cutesy, we need something (the conditional sampling diags is built around using fill-vals, but importantly, they are designed to passed around into other reductions ... so at a minimum, I will want to add the reduction diags and sampling diags to track fill vals manually). Let's discuss this week to see if we can figure out an intelligent solution or we need to keep plugging it along manually for now ... :/

**Comment by bartgol** [comment]
Funny you brought this up. Last week I spend a couple of hours seeing if I could "centralize" the handling of fill val, now that PR #7397 was merged. Let's chat offline.
